README.md
# phlib/path

[![Code Checks](https://img.shields.io/github/workflow/status/phlib/path/CodeChecks?logo=github)](https://github.com/phlib/path/actions/workflows/code-checks.yml)
[![Codecov](https://img.shields.io/codecov/c/github/phlib/path.svg?logo=codecov)](https://codecov.io/gh/phlib/path)
[![Latest Stable Version](https://img.shields.io/packagist/v/phlib/path.svg?logo=packagist)](https://packagist.org/packages/phlib/path)
[![Total Downloads](https://img.shields.io/packagist/dt/phlib/path.svg?logo=packagist)](https://packagist.org/packages/phlib/path)
![Licence](https://img.shields.io/github/license/phlib/path.svg)

PHP path handling component for dealing with escaped directory separators

## Install

Via Composer

``` bash
$ composer require phlib/path
```

## Usage

Creation of path instance

``` php
$path = \Phlib\Path::fromString('foo/bar/baz');
$info = $path->info(); // should return the same as `pathinfo`
```

Using path to parse paths with escaped directory separators 

``` php
$path = \Phlib\Path::fromString('foo/bar\\/baz');
echo $path->info(\Phlib\Path::INFO_BASENAME); // bar\/baz
```

Build paths with escaped separators

``` php
$parts = ['foo', 'bar/baz', 'taz'];
$path  = new \Phlib\Path($parts);
echo $path->toString(); // foo/bar\/baz/taz
```

## License

This package is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [2.1.0] - 2021-09-11
### Added
- Add support for PHP v8

## [2.0.0] - 2019-12-10
### Added
- Method parameter and return types. Implementations may need to be updated if
  they pass invalid values to Path methods.
### Removed
- **BC break**: Removed support for PHP v7.1 and v7.2 as they are no longer
[actively supported](https://php.net/supported-versions.php) by the PHP project

## [1.0.4] - 2019-12-09
### Added
- Add PHP version constraints to Composer config. No BC breaks as the constraint
  covers all current and past versions back to the implicit support for PHP v5.4.
  No future versions of PHP will be supported by *v1.x*.

## [1.0.3] - 2016-08-15
### Fixed
- PHP v5.4 compatibility for `trimEmptyParts()` by not using `array_filter()`.

## [1.0.2] - 2016-08-03
### Fixed
- Fix the logic for when to return a single value from `info()`.

## [1.0.1] - 2016-07-22
### Fixed
- Handle empty paths

## [1.0.0] - 2016-07-15
Initial Release


tests/PathTest.php
<?php

declare(strict_types=1);

namespace Phlib;

use PHPUnit\Framework\TestCase;

/**
 * @package Phlib\Path
 * @license LGPL-3.0
 */
class PathTest extends TestCase
{
    public function testCreateFromParts(): void
    {
        $parts     = ['foo', 'bar/baz', 'taz'];
        $phlibPath = new Path($parts);
        $this->assertEquals('foo/bar\\/baz/taz', $phlibPath->toString());
        // Also test magic `__toString()`
        $this->assertEquals('foo/bar\\/baz/taz', (string)$phlibPath);
    }

    /**
     * @dataProvider matchesPathInfoProvider
     * @param string $path
     */
    public function testMatchesPathInfo(string $path): void
    {
        $phpPathInfo   = pathinfo($path);
        $phlibPathInfo = Path::fromString($path)->info();

        $this->assertPathInfoEquals(
            $phpPathInfo,
            $phlibPathInfo,
            "Failed asserting path info matched for path '$path'"
        );
    }

    public function matchesPathInfoProvider(): array
    {
        return [
            ['foo'],
            ['foo/bar'],
            ['foo/bar/'],
            ['/foo'],
            ['/foo/bar'],
            ['/foo/bar/'],
            ['foo.baz'],
            ['foo/bar.baz'],
            ['/foo.baz'],
            ['/foo/bar.baz'],
            [''],
            ['/'],
            ['foo//bar'],
            ['/foo//bar'],
            ['/foo//bar/'],
            ['/foo//bar/'],
            ['/foo//bar//'],
            ['//foo//bar'],
            ['//foo//bar/'],
            ['//'],
            ['///'],
        ];
    }

    public function testEscapingPathSeparators(): void
    {
        $parts = [
            'foo/bar',
            'baz/taz\\',
            'boz/.woz'
        ];

        $parts = array_map(function ($part) {
            return Path::escapeName($part);
        }, $parts);

        $phlibPath = Path::fromString(implode('/', $parts));
        $this->assertEquals(3, count($phlibPath));
    }

    /**
     * @dataProvider ignoresEscapedSeparatorsProvider
     * @param string $path
     * @param int $expectedCount
     */
    public function testIgnoresEscapedSeparators(string $path, int $expectedCount): void
    {
        $phlibPath = Path::fromString($path);
        $this->assertEquals($expectedCount, count($phlibPath), "Failed asserting path count for path '$path'");
    }

    public function ignoresEscapedSeparatorsProvider(): array
    {
        return [
            ['foo\\/bar', 1],
            ['foo/bar\\/baz/foo', 3],
            ['foo/bar\\\\/baz/foo', 4],
            ['/foo/bar\\\\/baz/foo', 5],
            ['\\/foo/bar/baz', 3],
            ['foo/bar/\\/baz', 3],
            ['foo/bar/baz\\/', 3],
        ];
    }

    public function testPathInfoOption(): void
    {
        $path      = 'foo/bar/baz.taz';
        $phlibPath = Path::fromString($path);
        $this->assertEquals(pathinfo($path, PATHINFO_BASENAME), $phlibPath->info(Path::INFO_BASENAME));
        $this->assertEquals(pathinfo($path, PATHINFO_DIRNAME), $phlibPath->info(Path::INFO_DIRNAME));
        $this->assertEquals(pathinfo($path, PATHINFO_FILENAME), $phlibPath->info(Path::INFO_FILENAME));
        $this->assertEquals(pathinfo($path, PATHINFO_EXTENSION), $phlibPath->info(Path::INFO_EXTENSION));
    }

    public function testPathInfoMultipleOptions(): void
    {
        // a little extra behaviour which pathinfo can't handle
        $path      = 'foo/bar/baz.taz';
        $phlibPath = Path::fromString($path);

        $expected = [
            'filename'  => pathinfo($path, PATHINFO_FILENAME),
            'extension' => pathinfo($path, PATHINFO_EXTENSION),
        ];
        $actual = $phlibPath->info(Path::INFO_FILENAME | Path::INFO_EXTENSION);

        $this->assertPathInfoEquals($expected, $actual);
    }

    public function testPathInfoMultipleOptionsNotAllPresent(): void
    {
        $path = '';
        $phlibPath = Path::fromString($path);

        $expected = [
            'filename' => pathinfo($path, PATHINFO_FILENAME),
        ];
        $actual = $phlibPath->info(Path::INFO_FILENAME | Path::INFO_EXTENSION);

        $this->assertIsArray($actual);
        $this->assertPathInfoEquals($expected, $actual);
    }

    public function testCountable(): void
    {
        $phlibPath = Path::fromString('foo/bar/baz.taz');
        $this->assertEquals(3, count($phlibPath));
    }

    public function testOffsetAccess(): void
    {
        $phlibPath = Path::fromString('foo/bar/baz.taz');
        $this->assertTrue(isset($phlibPath[1]));
        $this->assertEquals('baz.taz', $phlibPath[2]);
    }

    public function testOffsetsAreImmutable(): void
    {
        $this->expectException(\RuntimeException::class);

        $phlibPath = Path::fromString('foo/bar/baz.taz');
        $phlibPath[1] = 'boo';
    }

    public function testOffsetsCannotUnset(): void
    {
        $this->expectException(\RuntimeException::class);

        $phlibPath = Path::fromString('foo/bar/baz.taz');
        unset($phlibPath[1]);
    }

    public function testIterable(): void
    {
        $phlibPath = Path::fromString('foo/bar/baz.taz');
        $out = [];
        foreach ($phlibPath as $part) {
            $out[] = $part;
        }
        $this->assertEquals('foo::bar::baz.taz', implode('::', $out));
    }

    public function testPartsAreUnescaped(): void
    {
        $name = 'my/file';
        $dir  = 'dir';
        $path = $dir . '/' . Path::escapeName($name);
        $phlibPath = Path::fromString($path);

        $this->assertEquals($name, $phlibPath->info(Path::INFO_BASENAME));
        $this->assertEquals($name, $phlibPath[1]);
    }

    public function testToStringReEscapes(): void
    {
        $name = 'my/file';
        $dir  = 'dir';
        $path = $dir . '/' . Path::escapeName($name);
        $phlibPath = Path::fromString($path);

        $this->assertEquals($path, $phlibPath->toString());
    }

    public function testDirnamePath(): void
    {
        $root = 'root';
        $dir  = 'dir';
        $path = "$root/$dir/file";

        $phlibPath = Path::fromString($path);
        $dirPath   = $phlibPath->getDirnamePath();

        $this->assertInstanceOf('\Phlib\Path', $dirPath);
        $this->assertEquals($dir, $dirPath->info(Path::INFO_BASENAME));
        $this->assertEquals($root, $dirPath->info(Path::INFO_DIRNAME));
    }

    public function testSlice(): void
    {
        $path = 'root/dir/file';
        $phlibPath = Path::fromString($path);

        $this->assertEquals('dir/file', $phlibPath->slice(1)->toString());
        $this->assertEquals('root/dir', $phlibPath->slice(0, -1)->toString());
        $this->assertEquals('file', $phlibPath->slice(-1)->toString());
        $this->assertEquals('root', $phlibPath->slice(0, 1)->toString());
    }

    public function testTrimStart(): void
    {
        $path = '/root/dir/file';
        $phlibPath = Path::fromString($path);

        $trimmed = $phlibPath->trimStart();

        $this->assertInstanceOf('\Phlib\Path', $trimmed);
        $this->assertEquals('root/dir/file', $trimmed->toString());
        $this->assertEquals(3, $trimmed->count());
    }

    /**
     * Helper method for comparing pathinfo arrays
     *
     * @param array $expected
     * @param array $actual
     * @param string $message
     */
    private function assertPathInfoEquals(array $expected, array $actual, string $message = ''): void
    {
        ksort($expected);
        ksort($actual);
        $this->assertEquals($expected, $actual, $message);
    }
}


src/Path.php
<?php

declare(strict_types=1);

namespace Phlib;

/**
 * @package Phlib\Path
 * @license LGPL-3.0
 */
class Path implements \ArrayAccess, \Countable, \IteratorAggregate
{
    /**
     * Constants for info keys, used for asking for info with the info method
     *
     * @see Path::info
     */
    public const INFO_DIRNAME   = 1;
    public const INFO_BASENAME  = 2;
    public const INFO_EXTENSION = 4;
    public const INFO_FILENAME  = 8;
    public const INFO_ALL       = 15;

    /**
     * The separator to use for delimiting parts of the path
     *
     * @var string $directorySeparator
     */
    private $directorySeparator;

    /**
     * The parts of the path
     *
     * @var string[] $parts
     */
    private $parts;

    /**
     * Array containing the info properties of the path
     *
     * @see Path::info
     * @var array info
     */
    private $info;

    /**
     * Mapping of info keys to the key names used in the output
     *
     * @see Path::info
     * @var array $infoKeys
     */
    private $infoKeys = [
        self::INFO_DIRNAME   => 'dirname',
        self::INFO_BASENAME  => 'basename',
        self::INFO_EXTENSION => 'extension',
        self::INFO_FILENAME  => 'filename',
    ];

    /**
     * Escape any directory separators which appear in a name (part of a path)
     *
     * This also escapes the escape character, to ensure that the escape character is still usable
     *
     * @param string $name
     * @param string $directorySeparator
     * @return string
     */
    public static function escapeName(string $name, string $directorySeparator = DIRECTORY_SEPARATOR): string
    {
        return strtr(
            $name,
            [
                '\\'                => '\\\\',                  // escape the escape character
                $directorySeparator  => "\\$directorySeparator",  // escape the separator
            ]
        );
    }

    /**
     * Unescape the directory separator character
     *
     * This generally shouldn't be needed outside of this class, as asking for a particular part of a path
     * from an instance of this class should return the unescaped name
     *
     * @param string $name
     * @param string $directorySeparator
     * @return string
     */
    public static function unescapeName(string $name, string $directorySeparator = DIRECTORY_SEPARATOR): string
    {
        return strtr(
            $name,
            [
                '\\\\'                  => '\\',
                "\\$directorySeparator" => "$directorySeparator",
            ]
        );
    }

    /**
     * Create a new path instance from a string path
     *
     * @param string $path
     * @param string $directorySeparator
     * @return Path
     */
    public static function fromString(string $path, string $directorySeparator = DIRECTORY_SEPARATOR): self
    {
        $parts = self::splitPath($path, $directorySeparator);
        return new self($parts, $directorySeparator);
    }

    /**
     * Splits a path by the directory separator
     *
     * This method could be as simple as an explode, were it not for needing to account for escaped
     * directory separators.
     *
     * @param string $path
     * @param string $directorySeparator
     * @return array
     */
    private static function splitPath(string $path, string $directorySeparator): array
    {
        $index    = 0;
        $length   = mb_strlen($path);
        $escaping = false;
        $out      = [];
        $lastSep  = 0;
        while ($index < $length) {
            $chr = mb_substr($path, $index, 1);
            if ($escaping) {
                $escaping = false;
            } elseif ($chr === '\\') {
                $escaping = true;
            } elseif ($chr === $directorySeparator) {
                $out[] = self::unescapeName(mb_substr($path, $lastSep, $index - $lastSep), $directorySeparator);
                $lastSep = $index + 1;
            }
            $index++;
        }
        $out[] = self::unescapeName(mb_substr($path, $lastSep, $index - $lastSep), $directorySeparator);
        return $out;
    }

    /**
     * Path constructor
     * @param array $parts
     * @param string $directorySeparator
     */
    public function __construct(array $parts, string $directorySeparator = DIRECTORY_SEPARATOR)
    {
        $parts = $this->trimEmptyParts($parts);

        $this->directorySeparator = $directorySeparator;
        $this->parts              = $parts;
    }

    /**
     * Get a new path instance for everything up to the parent directory
     *
     * @return Path
     */
    public function getDirnamePath(): self
    {
        return $this->slice(0, -1);
    }

    /**
     * Get a new path instance for a slice of the path (arguments should match array_slice)
     *
     * @param int $offset
     * @param int= $length
     * @return Path
     */
    public function slice(int $offset, int $length = null): self
    {
        $parts = $this->parts;
        $parts = isset($length) ? array_slice($parts, $offset, $length) : array_slice($parts, $offset);
        return new self($parts, $this->directorySeparator);
    }

    /**
     * Get a new path instance with any empty start part removed
     *
     * @return Path
     */
    public function trimStart(): self
    {
        $parts = $this->parts;
        if (count($parts) > 1 && empty($parts[0])) {
            array_shift($parts);
        }
        return new self($parts, $this->directorySeparator);
    }

    /**
     * Get a string representation of the path
     *
     * This will return the path with directory separators escaped for any part which contains them
     *
     * @return string
     */
    public function toString(): string
    {
        return implode($this->directorySeparator, $this->escapeParts($this->parts));
    }

    /**
     * Get the path info for this path
     *
     * Ignoring the escaping of directory separators, this method should return the same result as
     * pathinfo for any given path
     *
     * @see pathinfo
     * @param int $options
     * @return array|string
     */
    public function info(int $options = self::INFO_ALL)
    {
        $this->parseInfo();
        $info = [];
        foreach ($this->infoKeys as $key => $keyName) {
            if (isset($this->info[$key]) && $options & $key) {
                $info[$keyName] = $this->info[$key];
            }
        }
        if (($options & ($options - 1)) === 0) {
            // is a power of two, i.e. a single info option was specified
            return current($info);
        }
        return $info;
    }

    /**
     * Gets the info data for the info method if this has not already been done
     *
     * @see Path::info
     */
    private function parseInfo(): void
    {
        if (isset($this->info)) {
            return;
        }
        $this->info  = [];
        $parts       = $this->parts;
        if (empty($parts)) {
            $this->info = [
                self::INFO_BASENAME => '',
                self::INFO_FILENAME => '',
            ];
            return;
        }
        $basename    = array_pop($parts);

        $this->info[self::INFO_BASENAME] = $basename;
        if (empty($parts)) {
            $this->info[self::INFO_DIRNAME] = '.';
        } elseif (count($parts) === 1 && empty($parts[0])) {
            $this->info[self::INFO_DIRNAME] = '/';
        } else {
            $this->info[self::INFO_DIRNAME] = implode($this->directorySeparator, $this->escapeParts($parts));
        }

        if (($dotPos = strrpos($basename, '.')) !== false) {
            $this->info[self::INFO_FILENAME]  = substr($basename, 0, $dotPos);
            $this->info[self::INFO_EXTENSION] = substr($basename, $dotPos + 1);
        } else {
            $this->info[self::INFO_FILENAME]  = $basename;
        }
    }

    /**
     * Escapes each part in the given parts array to escape directory separators and returns the resulting array
     *
     * @param array $parts
     * @return array
     */
    private function escapeParts(array $parts): array
    {
        return array_map(function ($name) {
            return self::escapeName($name);
        }, $parts);
    }

    /**
     * Trims the empty elements from the parts array
     *
     * This has special behaviour to deal with empty paths, and paths which contain only a separator
     *
     * @param array $parts
     * @return array
     */
    private function trimEmptyParts(array $parts): array
    {
        $emptyLeading = empty($parts[0]);

        if (count($parts) === 1 && $emptyLeading) {
            // empty path (e.g. '')
            return [];
        }

        return array_filter($parts, function ($part, $index) use ($emptyLeading) {
            if ($index === 0) {
                // never trim first element to keep leading separators (e.g. '/foo')
                return true;
            }
            if ($index === 1 && $emptyLeading) {
                // never trim second element if first was empty (e.g. '/')
                return true;
            }
            return !empty($part);
        }, ARRAY_FILTER_USE_BOTH);
    }

    /**
     * @inheritdoc
     */
    public function count(): int
    {
        return count($this->parts);
    }

    /**
     * @inheritdoc
     */
    public function offsetExists($offset): bool
    {
        return array_key_exists($offset, $this->parts);
    }

    /**
     * @inheritdoc
     */
    public function offsetGet($offset)
    {
        return $this->parts[$offset];
    }

    /**
     * @inheritdoc
     */
    public function offsetSet($offset, $value): void
    {
        throw new \RuntimeException('Cannot modify parts of the path');
    }

    /**
     * @inheritdoc
     */
    public function offsetUnset($offset): void
    {
        throw new \RuntimeException('Cannot modify parts of the path');
    }

    /**
     * @inheritdoc
     */
    public function getIterator(): \Traversable
    {
        return new \ArrayIterator($this->parts);
    }

    public function __toString(): string
    {
        return $this->toString();
    }
}
